### PR TITLE
#160577613 User can rate an article

### PR DESCRIPTION
**Pull request by bevkololi**
## What does this PR do
Allows users to rate existing articles
Allows users to edit their previous article ratings
Create tests for the rating articles functionality

## Description of task to be completed.
Create rate articles endpoint
Create editing existing ratings endpoint

## Endpoints
| ENDPOINT                     | METHODS              | DESCRIPTION                                       |
|------------------------------|----------------------|---------------------------------------------------|
| /api/articles/{slug}/rate/              | POST,PUT            | Rate existing articles and edit previous article ratings          |

## Rating an article
The following is the sample of a request data
```
{
  "rating": {
    "rating": 4
  }
}
```

## Response
```
{
    "rating": 3,
    "rated_by": 2,
    "message": "You have successfully rated this article"
}
```

## Screenshots
screen shot 2018-10-08 at 17 59 59

When you get the article you rated you can see the average rating of all users on the article.

screen shot 2018-10-09 at 15 25 31

 The average rating will appear when you try to view the rated article

## Please Note
-Ratings should be within a range of 1 to 5
-You cannot rate your own article, therefore create a new user before rating

## Relevant Pivotal Tracker Stories
https://www.pivotaltracker.com/story/show/160577613

### CODE REVIEW

**Comment by gitaumoses4** [comment]
Kindly rebase against develop in order to resolve the conflicts. 
Otherwise, the task is well done.

**Review by charisschomba — COMMENTED** [review]
The PUT rating endpoint is crashing with the use of an incorrect slug.

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -129,3 +130,58 @@ def destroy(self, request, *args, **kwargs):
...
+
+    def post(self, request, *args, **kwargs):
+        """
+        Users can post article ratings
+        """
+        rating = request.data.get('rating', {})
+
+        try:
+            article = Article.objects.get(slug=kwargs['slug'])
+        except Exception:
```
Comment (esirK) [inline authors/apps/articles/views.py]: although this works, being more explicit would be better. Change this to except the exact Exception.

**Comment by bevkololi** [comment]
Okay, noted.

File: authors/apps/articles/tests/api/test_rating.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        self.assertIn(b"You have successfully rated this article", response.content)
+
+    def test_user_cannot_rate_unavailable_article(self):
+        """
+        If article does not exist, user cannot rate it
+        """
+        slug = None
+        rating = self.articleRating
+        self.register_and_login(self.rating_user)
+        response = self.client.post(reverse("articles:rate-article", kwargs={'slug': slug}), data=rating, format="json")
```
Comment (esirK) [inline authors/apps/articles/tests/api/test_rating.py:48]: Let the slug be a string like "None" to test the suggested bug by @charisschomba

**Comment by esirK** [comment]
Looks good. Work on the comments and fix the flake8 issues.

File: authors/apps/articles/tests/api/test_rating.py (hunk 2 above)
Comment (bevkololi) [reply authors/apps/articles/tests/api/test_rating.py:48]: I have updated this test. Thank you

File: authors/apps/articles/models.py [hunk 3]
```diff
@@ -69,5 +70,13 @@ def __str__(self):
...
 
 
-# register the pre_save signal
+class ArticleRating(models.Model):
+    """
+    Ratings that users give Articles
+    """
+    article = models.ForeignKey(Article, on_delete=models.CASCADE)
+    rating = models.IntegerField(default=0)
+    rated_by = models.ForeignKey(User, on_delete=models.CASCADE)
```
Comment (esirK) [inline authors/apps/articles/models.py:136]: You can provide a `related_name` field for the foreign key. Research why this might be important

**Comment by esirK** [comment]
Looks Good. Just remove this commit `0a0a1dd729012c7a0e33ede67a5e24c2a388df11` and work on the comment above and we will be good to merge.

File: authors/apps/articles/renderers.py [hunk 4]
```diff
@@ -53,3 +53,24 @@ def render(self, data, accepted_media_type=None, renderer_context=None):
                 'article': data
             }
         })
+
+
+class RatingJSONRenderer(JSONRenderer):
```
Comment (esirK) [inline authors/apps/articles/renderers.py]: You can use this class instead https://github.com/andela/ah-code-blooded/blob/develop/authors/apps/core/renderers.py

File: authors/apps/articles/models.py (hunk 3 above)
Comment (bevkololi) [reply authors/apps/articles/models.py:136]: Thanks I just did. https://stackoverflow.com/questions/2642613/what-is-related-name-used-for-in-django 👍

**Review by YAHYA-H — COMMENTED** [review]
Kindly add a link to your **RELEVANT PIVOTAL TRACKER ID**  😄 
Otherwise, task well done  👏

**Comment by esirK** [comment]
Looks Good.

**Pull request merged**